docs: review for Keycloak plugin (#277)
janus-idp/backstage-plugins: plugins/keycloak-backend/README.md

# Keycloak Backend Plugin for Backstage

> The Keycloak backend plugin integrates Keycloak into the Backstage.

Reviewer: ```suggestion
The Keycloak backend plugin integrates Keycloak into Backstage.
```

Author: Done

## Capabilities

The Keycloak backend plugin has the following capabilities:

- Synchronization of Keycloak users in a realm

> - Synchronization of Keycloak groups in a realm and their users

Reviewer: ```suggestion
- Synchronization of Keycloak groups, and users in those groups, in a realm
```

Author: I'm finding the suggestion a bit confusing. I think it's better to have bullets.

## Configuration

> The communication between the Backstage and Keycloak is enabled using Keycloak API. The support is available for using username/password or client credentials.

Reviewer: ```suggestion
Communication between Backstage and Keycloak is enabled by using the Keycloak API. Username / password or client credentials are both supported authentication methods.
```
Suggestion; I'd check with your dev team that this is the intended meaning, but otherwise it wasn't clear to me what support for username/password or client credentials was referring to.

Author: Here, the word support indicates that user either can use their username/password or can use client credentials, which are provided.

> The following table describes the parameters that you can configure in the `app-config.yaml` file to enable the plugin:

Reviewer: ```suggestion
You must configure the following parameters in the `app-config.yaml` file to enable the plugin:
```

Rewording suggestion, although I think it would be better to give an example yaml file here?

Reviewer: Also, should this section be formatted as a procedure rather than a reference doc?

Author: I think having a table reference is more user-friendly with all the setting values described. I'd keep it as-is.

| `username`     | Username to authenticate                                                                                                                        | ""            | Yes if using password based authentication           |
| `password`     | Password to authenticate                                                                                                                        | ""            | Yes if using password based authentication           |
| `clientId`     | Client Id to authenticate                                                                                                                       | ""            | Yes if using client credentials based authentication |
| `clientSecret` | Client Secret to authenticate                                                                                                                   | ""            | Yes if using client credentials based authentication |

When using client credentials, the access type must be set to `confidential` and service accounts must be enabled. You must add the following roles from the `realm-management` client role:

- `query-groups`
- `query-users`

> - `view-users`

Reviewer: Again, a YAML example here would be easier to follow

## Installation

1. Install the Backstage package into the backend. When not integrating with a published package, clone the repository locally and add the Backstage as follows:

   ```shell
   # From your Backstage root directory
   yarn add --cwd packages/backend @janus-idp/backstage-plugin-keycloak-backend
   ```

> 2. Add the following configuration to `app-config.yaml` file:

Reviewer: Just a tip for numbering in markdown; if you use `1.` for each numbered item, they should automatically generate the correct numbers. It's easier to maintain in the long run because e.g. if you remove a step or add one you don't have to manually update the numbering.
Example: https://stackoverflow.com/questions/12145232/create-an-automatically-numbered-list-in-markdown-format

Author: Thank you @abrennan89. 
However, we'll soon be converting the markdown files to asciidoc. If needed I'll make the change.

> 2. Add the following configuration to `app-config.yaml` file:

Reviewer: ```
Add the following configuration to the `app-config.yaml` file:
```

Author: Done

       ): Promise<Router> {
         const builder = await CatalogBuilder.create(env);

     +   builder.addEntityProvider(
     +     KeycloakOrgEntityProvider.fromConfig(env.config, {
     +       id: 'development',
     +       logger: env.logger,
     +       scheduler: env.scheduler,
     +     }),
     +   )
       ...
       }
     ```

   - Add a schedule directly inside the `packages/backend/src/plugins/catalog.ts` file as follows:

     ```diff
     // packages/backend/src/plugins/catalog.ts
     + import { KeycloakOrgEntityProvider } from '@janus-idp/backstage-plugin-keycloak-backend';

       export default async function createPlugin(
         env: PluginEnvironment,
       ): Promise<Router> {
         const builder = await CatalogBuilder.create(env);

     +   builder.addEntityProvider(
     +     KeycloakOrgEntityProvider.fromConfig(env.config, {
     +       id: 'development',
     +       logger: env.logger,
     +       schedule: env.scheduler.createScheduledTaskRunner({
     +         frequency: { minutes: 1 },
     +         timeout: { minutes: 1 },
     +         initialDelay: { seconds: 15 }
     +       }),
     +     }),
     +   )
       ...
       }

>      ```

Reviewer: I'd consider breaking this into a couple of steps instead, i.e. the required step of registering the plugin, and then the option step of configuring the scheduler

Author: This action is already in steps: required step is in numbered list and option steps are in bulleted list.

## Limitations

> If you have self-signed/corporate certificate issues, the temporary solution is to set the following environment variable before starting Backstage:

Reviewer: ```suggestion
If you have self-signed or corporate certificate issues, you can set the following environment variable before starting Backstage:
```

Author: Done

`NODE_TLS_REJECT_UNAUTHORIZED=0`

---

**NOTE**

The temporary solution of setting environment variable is not recommended.

> ---

Reviewer: Why not? If this isn't recommended, rather than having it be part of the main docs, I'd opt to add a troubleshooting docs section and discuss it there, including any github issues etc that can be linked that are related to this issue. If there are no plans to fix it or anything, I'm not sure it should be called out as temporary.

Author: Yeah makes sense. Currently, I've removed the word "temporary", and we can improve this section in the main docs. 
Changing the note as:

---

**NOTE**

The solution of setting the environment variable is not recommended.

---

### View imported users and groups in Backstage

> After configuring the plugin as described in the previous sections the plugin imports the users and groups on a regular basis when started. Once the first import is complete, select `User` to list the users from the catalog page:

Reviewer: ```suggestion
After you have successfully configured the plugin, it imports users and groups each time it is started.

After the first import is complete, select `User` to list the users from the catalog page:
```

Author: changing it to:

After configuring the plugin successfully, the plugin imports the users and groups each time when started. 

After the first import is complete, you can select `User` to list the users from the catalog page:
![](./images/users.jpg)

![](./images/users.jpg)

You can see the list of users on the page:
![](./images/user-list.jpg)

> When you select a user, you can see the information imported from the Keycloak:

Reviewer: ```suggestion
When you select a user, you can see the information imported from Keycloak:
```

Author: Done

![](./images/user2.jpg)

> Similarly, you can select a `Group`, view the list, and select or view the information imported from the

Reviewer: ```suggestion
You can also select a `Group`, view the list of groups, and select or view the information imported from 
```

Author: Done
